aten/src/ATen/native/cuda/UnaryGeometricKernels.cu

```diff
@@ -165,6 +166,67 @@ void tanh_kernel_cuda(TensorIterator& iter) {
   });
 }
 
+// We manually overload arccosh because std::acosh does not work with thrust::complex types.
+template<typename scalar_t>
+__host__ __device__ static inline scalar_t arccosh_wrapper(scalar_t v) {
+  return ::acosh(v);
+}
+
+template<typename T>
+__host__ __device__ static inline thrust::complex<T> arccosh_wrapper(thrust::complex<T> v) {
+  return thrust::acosh(v);
+}
+
+void arccosh_kernel_cuda(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND1(ScalarType::Half, iter.dtype(), "arccosh_cuda", [&]() {
+    using thrust_t = typename ztype_cuda<scalar_t>::thrust_t;
+    gpu_kernel(iter, []GPU_LAMBDA(thrust_t a) -> thrust_t {
+      return arccosh_wrapper(a);
+    });
+  });
+}
+
+// We manually overload arcsinh because std::asinh does not work with thrust::complex types.
+template<typename scalar_t>
+__host__ __device__ static inline scalar_t arcsinh_wrapper(scalar_t v) {
+  return ::asinh(v);
+}
+
+template<typename T>
+__host__ __device__ static inline thrust::complex<T> arcsinh_wrapper(thrust::complex<T> v) {
+  return thrust::asinh(v);
+}
+
+void arcsinh_kernel_cuda(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND1(ScalarType::Half, iter.dtype(), "arcsinh_cuda", [&]() {
+    using thrust_t = typename ztype_cuda<scalar_t>::thrust_t;
+    gpu_kernel(iter, []GPU_LAMBDA(thrust_t a) -> thrust_t {
+      return arcsinh_wrapper(a);
+    });
+  });
+}
+
+template<typename scalar_t>
+__host__ __device__ static inline scalar_t arctanh_wrapper(scalar_t v) {
+  return ::atanh(v);
+}
+
+template<typename T>
+__host__ __device__ static inline thrust::complex<T> arctanh_wrapper(thrust::complex<T> v) {
+  return thrust::atanh(v);
+}
+
+void arctanh_kernel_cuda(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(ScalarType::Half, ScalarType::BFloat16, iter.dtype(), "arctanh_cuda", [&]() {
```

Comment:
Out of curiosity, why is only this one supported for BFloat16?

Reply:
I think that's a valid question, and the only reason I did _not_ support for `arcsinh` and `arccosh` was that `sinh` and `cosh` didn't have that support while `tanh` did. I thought it's from the upstream, that support for BFloat16 wasn't desired/urgent. If you think it can be added, I can surely add support for BFloat16 to `arcsinh` and `arccosh`. Should have asked this before though, thank you for pointing this out.

Reply:
I am happy if the answer is: because sinh and cosh don't have it. I was just curious.

Reply:
So, for now, I added BFloat16 support for `acosh, asinh` as well. (Since I had added for `atanh` as well, so to preserve consistency - added for others as well unless and until there is any reason to not add for them). Added tests for this in `test/test_torch.py` as well to test for `_float_types2` (i.e. `torch.float, torch.float, torch.double, torch.bfloat16` if `TEST_WITH_ROCM` is set).

Please let me know if this should be reverted back. (https://github.com/pytorch/pytorch/pull/38388/commits/2f9f07a4344782e3a8f8009a4aa25229db8e37bc)